### PR TITLE
audit: online - limit download time based on HTTPS determination

### PR DESCRIPTION
**Pull request by JCount**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully run `brew tests` with your changes locally?

-----
Previously, the http_content_headers_and_checksum method always
downloaded the entire url, including headers. These downloads make
the audit more likely to hang or output false positives. The only
time the entire file is actually used is when verifying that the url
can use the HTTPS protocol instead of HTTP. Therefore, pass curl the
-I flag to only fetch headers unless the protocol is HTTP. If it is,
maintain the old behavior by passing --include to download both the
headers and the file.

### CODE REVIEW

File: Library/Homebrew/dev-cmd/audit.rb [hunk 1]
```diff
@@ -210,9 +211,9 @@ def self.check_http_content(url, user_agents: [:default])
     "The URL #{url} could use HTTPS rather than HTTP"
   end
 
-  def self.http_content_headers_and_checksum(url, user_agent: :default)
+  def self.http_content_headers_and_checksum(url, hash_needed, user_agent: :default)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/audit.rb]: `hash_needed: false`; boolean arguments are a bit evil.

File: Library/Homebrew/dev-cmd/audit.rb [hunk 2]
```diff
@@ -210,9 +211,9 @@ def self.check_http_content(url, user_agents: [:default])
...
     args = curl_args(
-      extra_args: ["--connect-timeout", "15", "--include", url],
+      extra_args: ["--connect-timeout", "15", hash_needed ? "--include" : "-I", url],
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/audit.rb]: Split this into a variable; the inline ternary operator is a bit hard to read.

File: Library/Homebrew/dev-cmd/audit.rb [hunk 3]
```diff
@@ -177,8 +177,9 @@ def self.check_http_content(url, user_agents: [:default])
 
     details = nil
     user_agent = nil
+    hash_needed = url.start_with?("http:") ? true : false
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/audit.rb]: `hash_needed = url.start_with?("http:")` will do the same thing.

**Comment by MikeMcQuaid** [comment]
> Previously, the http_content_headers_and_checksum method always
> downloaded the entire url, including headers. These downloads make
> the audit more likely to hang or output false positives. The only
> time the entire file is actually used is when verifying that the url
> can use the HTTPS protocol instead of HTTP. Therefore, pass curl the
> -I flag to only fetch headers unless the protocol is HTTP. If it is,
> maintain the old behavior by passing --include to download both the
> headers and the file.

🆒, makes sense. To avoid confusion I think the `details[:file_hash]` field shouldn't be set if the hash hasn't been queried.

Another option if you're feeling smart (for a future PR) would be to check the ETag rather than download hash if available. That would avoid the need for a full download.

File: Library/Homebrew/dev-cmd/audit.rb (hunk 3 above)
Comment (JCount) [inline Library/Homebrew/dev-cmd/audit.rb]: Yes, I am not quite certain why I wrote it the way I did. 🤔

**Comment by JCount** [comment]
We are in fact already parsing the ETag and checking them in addition to the hashes. Do you think that we should change the default behavior to simply comparing the ETags?

**Comment by MikeMcQuaid** [comment]
> We are in fact already parsing the ETags and checking them in addition to the hashes. Do you think that we should change the default behavior to simply comparing the ETags?

I'm not sure. Perhaps if the ETag comparison either fails or an ETag isn't present? What do you think? What does @broder think?

File: Library/Homebrew/dev-cmd/audit.rb [hunk 4]
```diff
@@ -210,9 +211,10 @@ def self.check_http_content(url, user_agents: [:default])
     "The URL #{url} could use HTTPS rather than HTTP"
   end
 
-  def self.http_content_headers_and_checksum(url, user_agent: :default)
+  def self.http_content_headers_and_checksum(url, hash_needed = false, user_agent: :default)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/audit.rb]: `hash_needed: false`

**Comment by MikeMcQuaid** [comment]
> Adding another condition to further prevent the need to download the entire file seems reasonable, but it may be best addressed in a separate PR. That way, we can 🚢 this PR, hopefully improving matters, and then iterate if need be.

@JCount 👍 from me; 👎 to scope creep.

File: Library/Homebrew/dev-cmd/audit.rb (hunk 4 above)
Comment (JCount) [inline Library/Homebrew/dev-cmd/audit.rb]: @MikeMcQuaid I originally tried that, but it causes `brew audit --online foo` to spit out:
```
Error: wrong number of arguments (2 for 1)
Please report this bug:
  http://docs.brew.sh/Troubleshooting.html
/usr/local/Homebrew/Library/Homebrew/dev-cmd/audit.rb:214:in `http_content_headers_and_checksum'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/audit.rb:182:in `block in check_http_content'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/audit.rb:181:in `each'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/audit.rb:181:in `check_http_content'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/audit.rb:1573:in `block in audit_urls'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/audit.rb:1565:in `each'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/audit.rb:1565:in `audit_urls'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/audit.rb:1314:in `audit'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/audit.rb:740:in `block in audit_specs'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/audit.rb:737:in `each'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/audit.rb:737:in `audit_specs'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/audit.rb:1259:in `audit'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/audit.rb:78:in `block in audit'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/audit.rb:74:in `each'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/audit.rb:74:in `audit'
/usr/local/Homebrew/Library/Homebrew/brew.rb:91:in `<main>'
```
Accidentally overlooking the scope creep, using `hash_needed = false` seemed to be the solution. However, I think I am overlooking something and cannot find an actual solution. Any guidance would be most welcome.

File: Library/Homebrew/dev-cmd/audit.rb [hunk 5]
```diff
@@ -210,9 +211,10 @@ def self.check_http_content(url, user_agents: [:default])
...
+    header_strategy = hash_needed ? "--include" : "-I"
```
Comment (broder) [inline Library/Homebrew/dev-cmd/audit.rb]: We should probably make "-I" into "--head" for clarity. Also, I'm not sure HEAD requests will work 100% of the time (e.g. on github). IIRC, the reason we're doing the full download every time is because it's more reliable.

**Comment by broder** [comment]
> Perhaps if the ETag comparison either fails or an ETag isn't present?

The fallback to the hash of the file was mainly for the case when the ETag isn't present.

File: Library/Homebrew/dev-cmd/audit.rb (hunk 5 above)
Comment (broder) [inline Library/Homebrew/dev-cmd/audit.rb]: You can also use `--range` but, again, I don't think it was reliable as some hosts didn't support it.

Comment (JCount) [inline Library/Homebrew/dev-cmd/audit.rb]: @broder Do you think it would be acceptable to simply use `--include --max-time 25` in the place of `--head` to provide more reliability? This would still eliminate the possibility of `audit --online` hanging on HTTPS urls, yet also address your concerns.

Comment (broder) [inline Library/Homebrew/dev-cmd/audit.rb]: It would almost certainly stop the hanging but wouldn't it change the results of `brew audit --online` (e.g. in the case of large files)? Is the hanging due to large files being downloaded or something else?

File: Library/Homebrew/dev-cmd/audit.rb (hunk 4 above)
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/audit.rb]: You'll need to use `hash_needed: false` in both the method definition and the callers of that method, too. Passing boolean arguments unless it's an accessor is generally not great because it's not as readable.

File: Library/Homebrew/dev-cmd/audit.rb (hunk 5 above)
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/audit.rb]: > We should probably make -I into --head for clarity.

Agreed.

**Comment by MikeMcQuaid** [comment]
> The fallback to the hash of the file was mainly for the case when the ETag isn't present.

Ok, cool. In that case I think we can (in the next PR if necessary) check just for the ETag first and then fall back to a hash only if it isn't available (or perhaps it doesn't match).

**Comment by ilovezfs** [comment]
Note that the hanging is not just about large files. It can get stuck and hang permanently on small or large files alike.

**Comment by MikeMcQuaid** [comment]
> Note that the hanging is not just about large files. It can get stuck and hang permanently on small or large files alike.

Ok. A max time/timeout seems sensible regardless then.

File: Library/Homebrew/dev-cmd/audit.rb (hunk 5 above)
Comment (JCount) [inline Library/Homebrew/dev-cmd/audit.rb]: Also note that this change in behavior only applies in situations where we were already only examining the headers and ignoring the rest of the output already.

File: Library/Homebrew/dev-cmd/audit.rb (hunk 4 above)
Comment (JCount) [inline Library/Homebrew/dev-cmd/audit.rb]: In that case, how about I switch to simply passing a string object?

File: Library/Homebrew/dev-cmd/audit.rb [hunk 6]
```diff
@@ -177,8 +177,9 @@ def self.check_http_content(url, user_agents: [:default])
 
     details = nil
     user_agent = nil
+    hash_needed = url.start_with?("http:")
     user_agents.each do |ua|
-      details = http_content_headers_and_checksum(url, user_agent: ua)
+      details = http_content_headers_and_checksum(url, hash_needed, user_agent: ua)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/audit.rb]: `details = http_content_headers_and_checksum(url, hash_needed: hash_needed, user_agent: ua)`

File: Library/Homebrew/dev-cmd/audit.rb [hunk 7]
```diff
@@ -188,11 +189,11 @@ def self.check_http_content(url, user_agents: [:default])
       return "The URL #{url} is not reachable (HTTP status code #{details[:status]})"
     end
 
-    return unless url.start_with? "http:"
+    return unless hash_needed
 
     secure_url = url.sub "http", "https"
     secure_details =
-      http_content_headers_and_checksum(secure_url, user_agent: user_agent)
+      http_content_headers_and_checksum(secure_url, true, user_agent: user_agent)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/audit.rb]: `http_content_headers_and_checksum(secure_url, hash_needed: true, user_agent: user_agent)`

File: Library/Homebrew/dev-cmd/audit.rb (hunk 4 above)
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/audit.rb]: `def self.http_content_headers_and_checksum(url, hash_needed: false, user_agent: :default)`

File: Library/Homebrew/dev-cmd/audit.rb (hunk 5 above)
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/audit.rb]: `header_strategy = hash_needed ? "--include" : "--head"`
